### ISSUE TITLE
how about enabling gradient calculation in topk() by default

### ISSUE THREAD

**Issue by zhaoyanpeng** [harvardnlp/pytorch-struct#55]
Pytorch enables grad inside `@lazy_property` methods by default. So in `@lazy_property` `argmax` we can use `torch.autograd.grad` without the risk of unrecorded computations.

In `topk` we are sure to use `torch.autograd.grad`. However, since it is not in lazy mode there is a chance that computations are not recorded. E.g., invoking `topk` in the scope of `torch.no_grad`. It is normal to disable grad during inference but it will implicitly affect the behavior of `topk`.

Would it be better to enable grad inside `topk` by default? The easiest way to achieve this is to modify
```
return self._struct(KMaxSemiring(k)).marginals(
    self.log_potentials, self.lengths, _raw=True
) 
```
to
```
with torch.enable_grad():
    return self._struct(KMaxSemiring(k)).marginals(
        self.log_potentials, self.lengths, _raw=True
    ) 
```
There could be different ways to achieve this. E.g., if we knew a function call is from `topk` we would switch on grad explicitly.

**Comment by srush** [comment]
Sure this seems fine.
